interfaces/prompting/errors.go

```diff
@@ -0,0 +1,198 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package prompting
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/snapcore/snapd/interfaces/prompting/errortypes"
+)
+
+var (
+	// NotFound errors when a prompt or rule is not found
+	ErrPromptNotFound = errors.New("cannot find prompt with the given ID for the given user")
+	ErrRuleNotFound   = errors.New("cannot find rule with the given ID")
+	ErrRuleNotAllowed = errors.New("user not allowed to request the rule with the given ID")
+
+	// Unsupported value errors, which are defined as functions below
+	ErrInvalidOutcome       = errInvalidOutcome
+	ErrInvalidLifespan      = errInvalidLifespan
+	ErrRuleLifespanSingle   = errRuleLifespanSingle
+	ErrInvalidInterface     = errInvalidInterface
+	ErrInvalidPermissions   = errInvalidPermissions
+	ErrPermissionsListEmpty = errPermissionsListEmpty
+	// Parse errors, which are defined as functions below
+	ErrInvalidDuration    = errInvalidDuration
+	ErrInvalidExpiration  = errInvalidExpiration
+	ErrInvalidPathPattern = errortypes.ErrInvalidPathPattern // avoid circular import in patterns package
+	// Validation errors, which should never be used directly apart from
+	// checking errors.Is(), and should otherwise always be wrapped in
+	// dedicated error types defined below.
+	ErrReplyNotMatchRequestedPath        = errors.New("path pattern in reply constraints does not match originally requested path")
+	ErrReplyNotMatchRequestedPermissions = errors.New("permissions in reply constraints do not include all requested permissions")
+	ErrRuleConflict                      = errors.New("a rule with conflicting path pattern and permission already exists in the rule database")
+
+	// Internal errors which are not handled specifically
+	ErrPromptsClosed      = errors.New("prompts backend has already been closed")
+	ErrRulesClosed        = errors.New("rules backend has already been closed")
+	ErrTooManyPrompts     = errors.New("cannot add new prompts, too many outstanding")
+	ErrRuleIDConflict     = errors.New("internal error: rule with conflicting ID already exists in the rule database")
+	ErrRuleDBInconsistent = errors.New("internal error: interfaces requests rule database left inconsistent")
+
+	// Errors which are used internally and should never be returned over the API
+	ErrNoMatchingRule          = errors.New("no rule matches the given path")
+	ErrInvalidID               = errors.New("invalid ID: format must be parsable as uint64")
+	ErrRuleExpirationInThePast = errors.New("cannot have expiration time in the past")
+)
+
+// Unsupported value errors, which are variants of errortypes.UnsupportedValueError
+
+func errInvalidOutcome(invalid OutcomeType) *errortypes.UnsupportedValueError {
```

Comment:
this should be probably be called (I|i)nvalidOutcomeErr(or), same for the others